lib/YuiRestClient.pm

```diff
@@ -17,110 +17,129 @@ use constant API_VERSION => 'v1';
+    my ($timeout) = @_;
+    $app->get_widget_controller()->set_timeout($timeout);
 }
 
-sub connect_to_app {
-    my $port = get_required_var('YUI_PORT');
-    my $host = get_required_var('YUI_SERVER');
-    die "Cannot set libyui REST API server" unless $host;
-    record_info('PORT',   "Used port for libyui: $port");
-    record_info('SERVER', "Connecting to: $host");
-    my $app = YuiRestClient::App->new({
+sub init_app {
+    my (%args)       = @_;
+    my $timeout      = $args{timeout}  || 10;
+    my $interval     = $args{interval} || 1;
+    my $installation = $args{installation};
+    $port = init_port();
+    $host = init_host($installation);
+
+    $app = YuiRestClient::App->new({
             port        => $port,
             host        => $host,
-            api_version => API_VERSION});
-    # As we start installer, REST API is not instantly available
-    $app->connect(timeout => 500, interval => 10);
-    set_app($app);
+            api_version => API_VERSION,
+            timeout     => $timeout,
+            interval    => $interval});
 }
 
-sub connect_to_running_app {
-    return get_app()->connect(timeout => 30, interval => 2);
+sub init_port {
+    return $port if defined $port;
+    $port = get_var('YUI_START_PORT', 39000);
+    $port += get_var('VNC') =~ /(?<vncport>\d+)/ ? $+{vncport} : int(rand(1000));
+    die "Cannot set port for YUI REST API" unless $port;
+
+    set_var('YUI_PORT', $port);
+    return $port;
 }
 
-sub setup_libyui_running_system {
-    zypper_call('in libyui-rest-api');
-
-    my $port = get_required_var('YUI_PORT');
-    my $host = get_required_var('YUI_SERVER');
-    record_info('PORT',   "Used port for libyui: $port");
-    record_info('SERVER', "Connecting to: $host");
-    set_var('YUI_PARAMS', get_yui_params_string());
-    # Add the port to permanent config and restart firewalld to apply the changes immediately.
-    # This is needed, because if firewall is restarted for some reason, then the port become
-    # closed (e.g. it was faced while saving settings in yast2 lan) and further tests will not
-    # be able to communicate with YaST modules.
-    assert_script_run("firewall-cmd --zone=public --add-port=$port/tcp --permanent");
-    assert_script_run('firewall-cmd --reload');
-    my $app = YuiRestClient::App->new({port => $port, host => $host, api_version => API_VERSION});
-    set_app($app);
+sub init_host {
+    return $host if defined $host;
+    my ($installation) = @_;
+    my $yuiport        = init_port();
+    my $ip_regexp      = qr/(?<ip>(\d+\.){3}\d+)/i;
+    if (check_var('BACKEND', 'qemu')) {
+        $host = 'localhost';
+    } elsif (is_pvm || is_ipmi) {
+        $host = get_var('SUT_IP');
+    } elsif (get_var('S390_ZKVM')) {
+        $host = get_var('VIRSH_GUEST');
+    } elsif (check_var('BACKEND', 's390x')) {
+        $installation ? select_console('install-shell') : select_console('root-console');
+        $host = YuiRestClient::Wait::wait_until(object => sub {
```

Comment:
I suggested the same in previous pull requests, we have another chance here to just save the sub in some variable and run it twice, so we don't repeat the code. Just in case you would like to do it now :)

Reply:
Done.